string.c

```diff
@@ -2500,32 +2500,39 @@ rb_check_lockedtmp(VALUE str)
     }
 }
 
+#define STR_UNMODIFIABLE_MASK (FL_FREEZE | STR_TMPLOCK | STR_CHILLED)
 static inline void
 str_modifiable(VALUE str)
 {
-    if (CHILLED_STRING_P(str)) {
-        CHILLED_STRING_MUTATED(str);
+    if (RB_UNLIKELY(FL_ANY_RAW(str, STR_UNMODIFIABLE_MASK))) {
+        if (CHILLED_STRING_P(str)) {
+            CHILLED_STRING_MUTATED(str);
+        }
+        rb_check_lockedtmp(str);
+        rb_check_frozen(str);
     }
-    rb_check_lockedtmp(str);
-    rb_check_frozen(str);
 }
 
 static inline int
 str_dependent_p(VALUE str)
 {
     if (STR_EMBED_P(str) || !FL_TEST(str, STR_SHARED|STR_NOFREE)) {
-        return 0;
+        return FALSE;
     }
     else {
-        return 1;
+        return TRUE;
     }
 }
 
+#define STR_DEPENDANT_MASK (STR_UNMODIFIABLE_MASK | STR_SHARED | STR_NOFREE)
 static inline int
 str_independent(VALUE str)
 {
-    str_modifiable(str);
-    return !str_dependent_p(str);
+    if (RB_UNLIKELY(FL_ANY_RAW(str, STR_DEPENDANT_MASK))) {
```

Comment:
This is missing the STR_EMBED_P check.

Reply:
Yeah, this is meant as a fast path, so we try to cheaply check the common case to bail out early.

And I don't think we can combine `STR_EMBED_P` in a single bitmask like this, so I just let it go in the "slowpath".  I'll add a comment.

Reply:
Makes sense to me now writing out the truth table:

```
original !str_dependent_p
embed shared nofree
0     0      0        = 1
0     0      1        = 0
0     1      0        = 0
0     1      1        = 0
1     0      0        = 1
1     0      1        = 1
1     1      1        = 1
```

when !shared && !nofree, it's always true.